src/server.rs

```diff
@@ -29,6 +29,38 @@ impl Server {
         })
     }
 
+    /// Creates a server configured to serve files from a given directory on
+    /// a given ip_address and a random port.
+    /// It is attempted to find a free port 16 times before giving up with `AddrInUse`.
+    /// On success the chosen port and the new `Server` instance are returned.
+    pub fn new_with_random_port<A: ToSocketAddrs, P: AsRef<Path>>(
+        ip_addr: &str,
+        serve_from: P,
+    ) -> Result<(u16, Self)> {
+        let mut rng = rand::thread_rng();
+        for _ in 0..16 {
+            let port: u16 = rng.gen_range(1025, u16::MAX);
+            let bind_to = format!("{}:{}", ip_addr, port);
+
+            // try binding to random port
+            match UdpSocket::bind(bind_to) {
+                // success
+                Ok(socket) => {
+                    let server = Self {
+                        socket,
+                        serve_dir: serve_from.as_ref().to_owned(),
+                    };
+                    return Ok((port, server));
+                }
+                // try again
+                Err(ref e) if e.kind() == io::ErrorKind::AddrInUse => continue,
+                // different error
+                Err(e) => return Err(e),
+            };
+        }
+        Err(io::ErrorKind::AddrInUse.into())
```

Comment:
```suggestion
        let port: u16 = rng.gen_range(1025, u16::MAX);
        let bind_to = format!("{}:{}", ip_addr, port);
        let socket = UdpSocket::bind(bind_to)?;
        
        let server = Self {
            socket,
            serve_dir: serve_from.as_ref().to_owned(),
        };
        
        Ok(socket)
```

I don't want to impose any policy over retrying for a socket. That can be the caller's responsibility if they get an Error from this function that indicates the address is already in use. I think this is more flexible for people that want to use this.